R/type-tibble.R

```diff
@@ -20,3 +20,30 @@ vec_ptype2.tbl_df.data.frame <- function(x, y, ..., x_arg = "x", y_arg = "y") {
 vec_ptype2.data.frame.tbl_df <- function(x, y, ..., x_arg = "x", y_arg = "y") {
   df_as_tibble(.Call(vctrs_type2_df_df, x, y, x_arg, y_arg))
 }
+
+
+tbl_ptype2.tbl_df <- function(x, y, ..., x_arg = "x", y_arg = "y") {
+  if (inherits_only(x, c("tbl_df", "tbl", "data.frame"))) {
+    UseMethod("tbl_ptype2.tbl_df", y)
+  } else {
+    stop_incompatible_type(x, y, x_arg = x_arg, y_arg = y_arg)
+  }
+}
+tbl_ptype2.tbl_df.default <- function(x, y, ..., x_arg = "x", y_arg = "y") {
+  stop_incompatible_type(x, y, x_arg = x_arg, y_arg = y_arg)
+}
+
+tbl_ptype2.tbl_df.data.frame <- function(x, y, ..., x_arg = "x", y_arg = "y") {
+  if (inherits_only(y, "data.frame")) {
```

Comment:
I find these a bit to hard to reason about because of `x` is checked in the generic, and the class of `y` is checked in the method.

Reply:
After writing the grouped-df and tbl-df methods, I think this makes it easier to write these methods by following a simple rule. Otherwise this might get too boiler-platy, making it easier to make a predicate mistake in one of the branches.